File: everything_parsing/06-ambiguity-disambiguation/java/ParseTree.java
import java.util.ArrayList;
import java.util.List;

/**
 * Module 05 — a parse tree, and the two canonical derivations you can read off it.
 *
 * A parse tree records WHICH productions built a string and HOW they nest — but
 * not the ORDER they were applied in. Two derivations of one tree:
 *
 *   leftmost   — at each step expand the leftmost not-yet-expanded nonterminal
 *   rightmost  — ... the rightmost one
 *
 * Both use the identical multiset of productions and end at the identical tree.
 * Only the intermediate sentential forms differ. That is the module's claim.
 */
public final class ParseTree {

    final String symbol;
    List<ParseTree> children;        // null => unexpanded; [] => expanded by an epsilon production
    Grammar.Production production;   // the production applied to expand this node; null if unexpanded

    ParseTree(String symbol) { this.symbol = symbol; }

    boolean isExpanded() { return children != null; }

    void expand(Grammar.Production p) {
        this.production = p;
        this.children = new ArrayList<>();
        for (String s : p.rhs()) children.add(new ParseTree(s));
    }

    // ─────────────────────────────────────────── build from production choices

    /**
     * Apply `choices` (production indices), each to the leftmost (or rightmost)
     * unexpanded nonterminal on the frontier. Returns the finished root.
     */
    static ParseTree build(Grammar g, int[] choices, boolean leftmost) {
        ParseTree root = new ParseTree(g.start);
        for (int c : choices) {
            ParseTree node = frontier(g, root, leftmost);
            if (node == null)
                throw new IllegalArgumentException("choices left over: tree already complete at choice " + c);
            Grammar.Production p = g.productions.get(c);
            if (!p.lhs().equals(node.symbol))
                throw new IllegalArgumentException(
                    "production " + c + " (" + p + ") does not match frontier nonterminal '" + node.symbol + "'");
            node.expand(p);
        }
        if (frontier(g, root, leftmost) != null)
            throw new IllegalArgumentException("choices ran out before the tree was complete");
        return root;
    }

    /** The leftmost (or rightmost) unexpanded nonterminal, or null if the tree is done. */
    private static ParseTree frontier(Grammar g, ParseTree root, boolean leftmost) {
        List<ParseTree> fr = new ArrayList<>();
        collect(g, root, fr);
        if (fr.isEmpty()) return null;
        return leftmost ? fr.get(0) : fr.get(fr.size() - 1);
    }
    private static void collect(Grammar g, ParseTree n, List<ParseTree> out) {
        if (!n.isExpanded()) {
            if (g.isNonterminal(n.symbol)) out.add(n);
            return;
        }
        for (ParseTree k : n.children) collect(g, k, out);
    }

    // ─────────────────────────────────────────── read a derivation off a finished tree

    record Step(List<String> form, Grammar.Production applied) {}

    static List<Step> derivation(Grammar g, ParseTree root, boolean leftmost) {
        List<Step> steps = new ArrayList<>();
        List<ParseTree> frontierNodes = new ArrayList<>(List.of(root));
        steps.add(new Step(symbols(frontierNodes), null));

        while (true) {
            int i = pick(g, frontierNodes, leftmost);
            if (i < 0) break;
            ParseTree node = frontierNodes.remove(i);
            frontierNodes.addAll(i, node.children);
            steps.add(new Step(symbols(frontierNodes), node.production));
        }
        return steps;
    }

    private static int pick(Grammar g, List<ParseTree> nodes, boolean leftmost) {
        if (leftmost) {
            for (int i = 0; i < nodes.size(); i++)
                if (isExpandedNonterminal(g, nodes.get(i))) return i;
        } else {
            for (int i = nodes.size() - 1; i >= 0; i--)
                if (isExpandedNonterminal(g, nodes.get(i))) return i;
        }
        return -1;
    }
    private static boolean isExpandedNonterminal(Grammar g, ParseTree n) {
        return n.isExpanded() && g.isNonterminal(n.symbol);
    }

    private static List<String> symbols(List<ParseTree> nodes) {
        List<String> s = new ArrayList<>();
        for (ParseTree n : nodes) s.add(n.symbol);
        return s;
    }

    // ─────────────────────────────────────────── yield + rendering

    /** The terminal frontier, left to right — the string the tree parses. */
    List<String> terminalYield() {
        List<String> out = new ArrayList<>();
        yieldInto(this, out);
        return out;
    }
    private static void yieldInto(ParseTree n, List<String> out) {
        if (!n.isExpanded()) { out.add(n.symbol); return; }
        for (ParseTree k : n.children) yieldInto(k, out);   // epsilon production: no children, adds nothing
    }

    /** Multiset of productions used, sorted by production index, as text. */
    static String productionMultiset(ParseTree root) {
        List<Grammar.Production> used = new ArrayList<>();
        collectProductions(root, used);
        used.sort((x, y) -> Integer.compare(x.index(), y.index()));
        List<String> parts = new ArrayList<>();
        for (Grammar.Production p : used) parts.add(compact(p));
        return "[" + String.join(", ", parts) + "]";
    }
    private static void collectProductions(ParseTree n, List<Grammar.Production> out) {
        if (!n.isExpanded()) return;
        out.add(n.production);
        for (ParseTree k : n.children) collectProductions(k, out);
    }
    private static String compact(Grammar.Production p) {
        return p.lhs() + "->" + (p.rhs().isEmpty() ? "eps" : String.join("", p.rhs()));
    }

    /** ASCII tree, `+-` branches, `|  ` / `   ` continuations. */
    String render() {
        StringBuilder sb = new StringBuilder(symbol).append('\n');
        renderKids(this, "", sb);
        return sb.toString().stripTrailing();
    }
    private static void renderKids(ParseTree n, String prefix, StringBuilder sb) {
        if (!n.isExpanded()) return;
        if (n.children.isEmpty()) { sb.append(prefix).append("+- epsilon\n"); return; }
        for (int i = 0; i < n.children.size(); i++) {
            ParseTree k = n.children.get(i);
            boolean last = i == n.children.size() - 1;
            sb.append(prefix).append("+- ").append(k.symbol).append('\n');
            renderKids(k, prefix + (last ? "   " : "|  "), sb);
        }
    }
}

File: everything_parsing/06-ambiguity-disambiguation/js/parsetree.mjs
// Module 05 — a parse tree, and the two canonical derivations you can read off it.
//
//   leftmost   — at each step expand the leftmost not-yet-expanded nonterminal
//   rightmost  — ... the rightmost one
//
// Both use the identical multiset of productions and end at the identical tree.
// Only the intermediate sentential forms differ.

import { productionToString } from "./grammar.mjs";

export class ParseTree {
  constructor(symbol) {
    this.symbol = symbol;
    this.children = null;      // null => unexpanded; [] => expanded by an epsilon production
    this.production = null;    // the production applied to expand this node
  }

  get isExpanded() { return this.children !== null; }

  expand(p) {
    this.production = p;
    this.children = p.rhs.map(s => new ParseTree(s));
  }

  // ── build from production choices ──

  static build(g, choices, leftmost) {
    const root = new ParseTree(g.start);
    for (const c of choices) {
      const node = frontier(g, root, leftmost);
      if (!node) throw new Error(`choices left over: tree already complete at choice ${c}`);
      const p = g.productions[c];
      if (p.lhs !== node.symbol)
        throw new Error(
          `production ${c} (${productionToString(p)}) does not match frontier nonterminal '${node.symbol}'`);
      node.expand(p);
    }
    if (frontier(g, root, leftmost)) throw new Error("choices ran out before the tree was complete");
    return root;
  }

  // ── read a derivation off a finished tree ──

  static derivation(g, root, leftmost) {
    const steps = [];
    let frontierNodes = [root];
    steps.push({ form: frontierNodes.map(n => n.symbol), applied: null });

    for (;;) {
      const i = pick(g, frontierNodes, leftmost);
      if (i < 0) break;
      const node = frontierNodes[i];
      frontierNodes = [...frontierNodes.slice(0, i), ...node.children, ...frontierNodes.slice(i + 1)];
      steps.push({ form: frontierNodes.map(n => n.symbol), applied: node.production });
    }
    return steps;
  }

  // ── yield + rendering ──

  terminalYield() {
    const out = [];
    (function walk(n) {
      if (!n.isExpanded) { out.push(n.symbol); return; }
      n.children.forEach(walk);           // epsilon production: no children, adds nothing
    })(this);
    return out;
  }

  render() {
    let sb = this.symbol + "\n";
    const kids = (n, prefix) => {
      if (!n.isExpanded) return;
      if (n.children.length === 0) { sb += prefix + "+- epsilon\n"; return; }
      n.children.forEach((k, i) => {
        const last = i === n.children.length - 1;
        sb += prefix + "+- " + k.symbol + "\n";
        kids(k, prefix + (last ? "   " : "|  "));
      });
    };
    kids(this, "");
    return sb.replace(/\s+$/, "");
  }
}

export function productionMultiset(root) {
  const used = [];
  (function walk(n) {
    if (!n.isExpanded) return;
    used.push(n.production);
    n.children.forEach(walk);
  })(root);
  used.sort((a, b) => a.index - b.index);
  return "[" + used.map(compact).join(", ") + "]";
}

export function compact(p) {
  return p.lhs + "->" + (p.rhs.length === 0 ? "eps" : p.rhs.join(""));
}

// ── helpers ──

function frontier(g, root, leftmost) {
  const fr = [];
  (function collect(n) {
    if (!n.isExpanded) { if (g.isNonterminal(n.symbol)) fr.push(n); return; }
    n.children.forEach(collect);
  })(root);
  if (fr.length === 0) return null;
  return leftmost ? fr[0] : fr[fr.length - 1];
}

function pick(g, nodes, leftmost) {
  const ok = n => n.isExpanded && g.isNonterminal(n.symbol);
  if (leftmost) {
    for (let i = 0; i < nodes.length; i++) if (ok(nodes[i])) return i;
  } else {
    for (let i = nodes.length - 1; i >= 0; i--) if (ok(nodes[i])) return i;
  }
  return -1;
}
